src/SimpleInjector.Tests.Unit/InstanceProducerTests.cs

```diff
@@ -136,6 +136,40 @@ public void VisualizeObjectGraph_NestedObjectGraphWithMultipleDecorators_BuildsT
             Assert.AreEqual(expectedObjectGraph, actualObjectGraph);
         }
 
+        [TestMethod]
+        public void VisualizeObjectGraph_WithLifetimeInformation_BuildsTheExpectedGraph()
+        {
+            // Arrange
+            string expectedObjectGraph =
+@"PluginDecorator( // Transient
+    PluginDecoratorWithDependencyOfType<FakeTimeProvider>( // Transient
+        FakeTimeProvider( // Transient),
```

Comment:
How do you want the comma in this example with the FakeTimeProvider?
```
PluginDecorator( // Transient
    PluginDecoratorWithDependencyOfType<FakeTimeProvider>( // Transient
        FakeTimeProvider() // Transient,
        PluginDecorator<Int32>( // Transient
            PluginWithDependencyOfType<RealTimeProvider>( // Transient
                RealTimeProvider() // Transient
            )
        )
    )
)
```
the comma (,) before the comment (//) ? If you would like to be closer to valid C# then the comma should be before the comment tag.

Reply:
Like this:

```
PluginDecorator( // Transient
    PluginDecoratorWithDependencyOfType<FakeTimeProvider>( // Transient
        FakeTimeProvider(), // Transient
        PluginDecorator<Int32>( // Transient
            PluginWithDependencyOfType<RealTimeProvider>( // Transient
                RealTimeProvider())))) // Transient
```

That would still not be valid C#, but it would if we would just add the `new` statements:

``` c#
new PluginDecorator( // Transient
    new PluginDecoratorWithDependencyOfType<FakeTimeProvider>( // Transient
        new FakeTimeProvider(), // Transient
        new PluginDecorator<Int32>( // Transient
            new PluginWithDependencyOfType<RealTimeProvider>( // Transient
                new RealTimeProvider())))) // Transient
```

The idea is to make it as readable as possible to C# developers. However, adding the `new` keywords would add noise, and is misleading, because a `Singletons` are not newed up like this, but cached. The reality is obviously more complex, and this just shows the structure, of course.

Reply:
crystal clear :) I misunderstood you the first time, so I deleted my comment. 

I try to pick it up tomorrow/asap :)

Reply:
Fixed this. I had to rewrite the existing method with linq query statement with a recursive method and StringBuilder to support various scenario's like closing parenthesis or komma. 

Another thing is when the instance got dependencies, the closing parenthesis will always be on a newline.

Reply:
How difficult is it to keep the closing brackets on the same line as they were before (but with the lifestyle comment after it)? But perhaps I'm asking something that is extremely difficult to achieve?

Reply:
@dotnetjunkie last night I was pondering for a fix for this. It's do-able but the code isn't really readable/clean code anymore. In my pull request I'm providing a StringBuilder to the recursive method to write the output.

What do you think of this outputs (inspired by cyclic dependency graph output):
```csharp
PluginDecorator( /* Transient */
    PluginDecoratorWithDependencyOfType<FakeTimeProvider>( /* Transient */
        FakeTimeProvider( /* Transient */ ), 
        PluginDecorator<Int32>( /* Transient */
            PluginWithDependencyOfType<RealTimeProvider>( /* Transient */
                RealTimeProvider( /* Transient */ ))))) 
```

When cyclic denpendency graph detected:
```csharp
InstanceProducerTests.NodeFactory( /* Transient */
    IEnumerable<InstanceProducerTests.INode>( /* Transient */
        InstanceProducerTests.NodeFactory( /* Transient */
            IEnumerable<InstanceProducerTests.INode>(/* cyclic dependency graph detected */))))
```

This  is also "valid" C# code. I think we can keep the current implementation with linq query syntax instead my pull request. I'm looking forward to your input!

Reply:
What if you extract the building of the object graph into a new type, let's say `ObjectGraphBuilder`, wrapping a `StringBuilder`. That would probably work. For instance:

``` c#
    internal sealed class ObjectGraphBuilder
    {
        private readonly StringBuilder builder = new StringBuilder(100);
        private readonly Stack<ProducerEntry> producers = new Stack<ProducerEntry>();
        private readonly bool writeLifestyles;

        private ProducerEntry stillToWriteLifestyleEntry;
        private int indentingDepth;
        
        public ObjectGraphBuilder(bool writeLifestyles)
        {
            this.writeLifestyles = writeLifestyles;
        }
        
        public override string ToString() => this.builder.ToString();

        internal void BeginInstanceProducer(InstanceProducer producer)
        {
            if (this.producers.Any())
            {
                this.AppendLifestyle(this.producers.Peek());
                this.AppendNewLine();
            }

            this.producers.Push(new ProducerEntry(producer));

            this.Append(producer.ImplementationType.ToFriendlyName());
            this.Append("(");

            this.indentingDepth++;
        }

        internal void AppendCyclicInstanceProducer(InstanceProducer producer, bool last)
        {
            this.BeginInstanceProducer(producer);
            this.Append("/* cyclic dependency graph detected */");
            this.EndInstanceProducer(last);
        }

        private void AppendNewLine()
        {
            this.Append(Environment.NewLine);
            this.AppendIndent();
        }

        internal void EndInstanceProducer(bool last)
        {
            var entry = this.producers.Pop();

            this.indentingDepth--;

            this.Append(")");

            if (!last)
            {
                this.Append(",");

                if (this.stillToWriteLifestyleEntry != null)
                {
                    this.AppendLifestyle(this.stillToWriteLifestyleEntry);
                    this.stillToWriteLifestyleEntry = null;
                }
            }

            if (!entry.LifestyleWritten)
            {
                this.stillToWriteLifestyleEntry = entry;
            }

            if (!this.producers.Any())
            {
                this.AppendLifestyle(this.stillToWriteLifestyleEntry);
            }
        }

        private void AppendLifestyle(ProducerEntry entry)
        {
            if (this.writeLifestyles && !entry.LifestyleWritten)
            {
                this.Append(" // ");
                this.Append(entry.Producer.Lifestyle.Name);
                entry.LifestyleWritten = true;
            }
        }

        private void AppendIndent()
        {
            for (int i = 0; i < indentingDepth * 4; i++)
            {
                this.Append(" ");
            }
        }

        private void Append(string value) => this.builder.Append(value);

        private sealed class ProducerEntry
        {
            public ProducerEntry(InstanceProducer producer)
            {
                this.Producer = producer;
            }

            public InstanceProducer Producer { get; }
            public bool LifestyleWritten { get; set; }
        }
    }
```

I agree, it's not the cleanest code, but I do really like to have the lifestyles at the end (sorry for nagging about this).

Would that work?

Reply:
No apologies needed. You guys are the maintainers 👍 Anyway I'll take a look at it.

Reply:
@dotnetjunkie the ObjectGraphBuilder definitely worked. I also found a wrong assertion in my unittest. See my latest changes/commit on the pull request